### PR TITLE
UI support for 0.0.10

### PR DESCRIPTION
**Pull request by zathras-crypto**
Note: Not yet final.
Few quirks still remaining - eg signalling is no longer functioning properly and default column sizes on balances tab look off, still looking into these bits.  I think I've resolved most issues affecting functionality though :)

### CODE REVIEW

**Comment by dexX7** [comment]
To clarify: is there going to be a 0.10 based release without meta DEx?

**Comment by zathras-crypto** [comment]
> To clarify: is there going to be a 0.10 based release without meta DEx?

Nope.  MetaDEx is the next major feature release and 0.0.10 won't drop until it's ready.  I think @msgilligan is going to be taking a look at getting the MetaDEx tests into jenkins but I know he's busy with OmniJ at the mo.

**Comment by dexX7** [comment]
> MetaDEx is the next **major** feature release

This was actually the reason why I bumped the version to 0.0.9.2 at first. I didn't test the UI integration yet, but this seems like the biggest chuck to overcome, and it looks like most of it is almost done, so I'd assume a 0.0.9.1-rel-but-based-on-bitcoin-core-0.10 version could be ready within a short time. Is there a reason to skip a minor release, based on Bitcoin Core 0.10?

I'm actually asking, and I'll probably start a new thread for this later, because the meta DEx parts are  disabled, and depending on whether there is a minor release or not, they could be enabled (as well as the other meta DEx code parts), or they better remain disabled. :)

**Comment by zathras-crypto** [comment]
There are still some bits to the UI I need to finish (I just wanted to get my local stuff up to GitHub) but yep it could be adapted to a 0.0.9.2 release in a short time.  I guess I tend to look at it the other way, why go to the distraction for a minor release?  Are we sure the migration to Bitcoin 0.10 is solid? 

I'm about to dive into the MetaDEx stuff once I'm finished up with the UI, first part of that is the MetaDEx UI - you'll notice I've already added the Exchange tab back in (though currently blank).

> because the meta DEx parts are disabled

Admittedly have been focused on other stuff and haven't looked at the MetaDEx stuff for months, but for testnet:
```
zathras@coredev01:~/github/build/omnicore$ src/bitcoin-cli getinfo_MP | grep trades
    "totaltrades" : 102,
```
I had thought it was enabled (testnet only) - I'll have to take a look into that - can you elaborate on which bits you see as disabled?

**Comment by dexX7** [comment]
> I guess I tend to look at it the other way, why go to the distraction for a minor release?

Imho this is a milestone worth to publish, and it's "on the way" anyway, so I guess the only overhead would be to build binaries and tag a release. Nearly 40 % of all nodes on bitnodes.io are using 0.10, which seems rather significant.

> I'm about to dive into the MetaDEx stuff once I'm finished up with the UI

Yup, no rush. For example the RPC calls are disabled at the moment.

File: src/qt/lookupspdialog.cpp [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+LookupSPDialog::LookupSPDialog(QWidget *parent) :
+    QDialog(parent),
+    ui(new Ui::LookupSPDialog),
+    model(0)
+{
+    ui->setupUi(this);
+    this->model = model;
+
+    // populate placeholder text
+    ui->searchLineEdit->setPlaceholderText("ID, name or issuer");
```
Comment (dexX7) [inline src/qt/lookupspdialog.cpp]:
```c++
#if QT_VERSION >= 0x040700
    ui->searchLineEdit->setPlaceholderText("ID, name or issuer");
#endif
```

File: src/qt/metadexdialog.cpp [hunk 2]
```diff
@@ -0,0 +1,666 @@
...
+    ui->setupUi(this);
+    this->model = model;
+    //open
+    global_metadex_market = 3;
+
+    //prep lists
+    ui->buyList->setColumnCount(3);
+    ui->sellList->setColumnCount(3);
+    ui->buyList->verticalHeader()->setVisible(false);
+    ui->buyList->horizontalHeader()->setResizeMode(QHeaderView::Stretch);
```
Comment (dexX7) [inline src/qt/metadexdialog.cpp]: ´´´c++
#if QT_VERSION < 0x050000
    ui->buyList->horizontalHeader()->setResizeMode(QHeaderView::Stretch);
#else
    ui->buyList->horizontalHeader()->setSectionResizeMode(QHeaderView::Stretch);
#endif
```

File: src/qt/metadexdialog.cpp [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+    ui->buyList->setShowGrid(false);
+    ui->buyList->setSelectionBehavior(QAbstractItemView::SelectRows);
+    ui->buyList->setEditTriggers(QAbstractItemView::NoEditTriggers);
+    ui->buyList->setSelectionMode(QAbstractItemView::SingleSelection);
+    ui->buyList->setAlternatingRowColors(true);
+    ui->sellList->verticalHeader()->setVisible(false);
+    ui->sellList->horizontalHeader()->setResizeMode(QHeaderView::Stretch);
```
Comment (dexX7) [inline src/qt/metadexdialog.cpp]: ´´´c++
#if QT_VERSION < 0x050000
    ui->sellList->horizontalHeader()->setResizeMode(QHeaderView::Stretch);
#else
    ui->sellList->horizontalHeader()->setSectionResizeMode(QHeaderView::Stretch);
#endif
´´´

File: src/qt/orderhistorydialog.cpp [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+    ui->orderHistoryTable->setHorizontalHeaderItem(1, new QTableWidgetItem("Date"));
+    ui->orderHistoryTable->setHorizontalHeaderItem(2, new QTableWidgetItem("Status"));
+    ui->orderHistoryTable->setHorizontalHeaderItem(3, new QTableWidgetItem("Trade Details"));
+    ui->orderHistoryTable->setHorizontalHeaderItem(4, new QTableWidgetItem("Sold"));
+    ui->orderHistoryTable->setHorizontalHeaderItem(5, new QTableWidgetItem("Received"));
+    ui->orderHistoryTable->verticalHeader()->setVisible(false);
+    ui->orderHistoryTable->setSelectionBehavior(QAbstractItemView::SelectRows);
+    ui->orderHistoryTable->setEditTriggers(QAbstractItemView::NoEditTriggers);
+    ui->orderHistoryTable->setSelectionMode(QAbstractItemView::SingleSelection);
+    ui->orderHistoryTable->horizontalHeader()->setResizeMode(3, QHeaderView::Stretch);
```
Comment (dexX7) [inline src/qt/orderhistorydialog.cpp]:
```c++
#if QT_VERSION < 0x050000
    ui->orderHistoryTable->horizontalHeader()->setResizeMode(3, QHeaderView::Stretch);
#else
    ui->orderHistoryTable->horizontalHeader()->setSectionResizeMode(3, QHeaderView::Stretch);
#endif
```

**Comment by dexX7** [comment]
@zathras-crypto: something similar happened earlier: https://bitcointalk.org/index.php?topic=808597.0;all

**Comment by zathras-crypto** [comment]
I noticed on the thread about pricing in MSC - all the examples you linked had the unit prices columns placed differently:
image
Do you think we should make that change?

**Comment by dexX7** [comment]
> Do you think we should make that change?

Yes, I think this would be helpful, to compare the prices. And on the first glimpse, I'd swap left and right, bascially to end up with something similar to the mockup I posted.

Ideally, the values left would be right aligned, as in your picture, but the values at the right should be left aligned, to bring the numbers closer together (just picked up the picture, for the general idea):

ed20510a-6b8b-11e4-9a97-69da7d6f87ec

**Comment by dexX7** [comment]
@zathras-crypto: I was thinking about the class C integration and the situation, when we merged it.

I suggest the following: the use of the green `"ready for review"` label shall indicate, whether the author thinks a PR is ready to be merged. This makes asynchronous collaboration easier, because then we can skip the step of reassuring, whether it's really ready and could be merged.

Further, `"work in progress"` shall indicate: "there is currently no need to keep a close eye on this, and I'll update, once the work is finished, or I need a second opinion".

The `"request for comments"` label could be used to signal: "there is something to discuss/let's reflect on it's current state".

If you have something else in mind, this could of course be extended. :)

**Comment by dexX7** [comment]
@zathras-crypto: even though this is still WIP, is there any reason not to consolidate at this point?

**Comment by zathras-crypto** [comment]
> is there any reason not to consolidate at this point?

By consolidate do you mean merge mate?

**Comment by dexX7** [comment]
@zathras-crypto: sorry, yes, this is what I mean. :)

**Comment by dexX7** [comment]
A few words to the motivation:

Over the last days I noticed that I started a new branch, based on the UI branch, and then ported the changes back to the non-UI branch to prepare a PR, which is sort of telling.

Of course this does ot mean the UI should be considered as finalized, but if there is no strong reason not to merge at this point, e.g. because it breaks things, or has very unexpected behavior, then I think it would be beneficial.

**Comment by zathras-crypto** [comment]
> A few words to the motivation

Sure - I'm happy to merge if you are mate - 0.0.10 is a WIP and shouldn't be used outside of testing at the moment anyway so given that I've no objections to merging...

**Comment by zathras-crypto** [comment]
> Which data is manipulated in Map, Data or Table?

Brief summary of how it works:
* ```PopulateTradeHistoryMap()``` is responsible for filling up ```tradeHistoryMap``` with  ```TradeHistoryObject```'s from new pending or wallet transactions.  This map is then the source of data for the table in the UI
* ```UpdateTradeHistoryTable()``` is responsible for using the data in the map to add new rows to the table in the UI
* ```UpdateData()``` is responsible for updating status and amounts in trades that already exist in the map

So, in a nutshell:
1) a new transaction (pending or wallet) is detected and added to the ```tradeHistoryMap```
2) this new entry in the map is added to the UI table by ```UpdateTradeHistoryTable()```
3) when new blocks come in if the status or amounts have changed the row is modified by ```UpdateData()```

Essentially it means the transaction is only parsed once in 1) rather than repeatedly.

> Maybe this could be renamed to PopulateTradeHistory(), UpdateTradeStatus(), UpdateConfirmationStatus().

I thought the names were fairly descriptive already - maybe perhaps ```UpdateData()``` could be renamed...  

Does that help provide some context at all?

**Comment by dexX7** [comment]
You're the best, thanks. I love the very detailed descriptions! :)

After reading the code, it was pretty clear what's happening, but when seeing only `Populate_Map()`, `Update_Table()` and `UpdateData()`, then one might guess that data is manipulated somehow, but it's not very obvious what happens, thus the suggestion for more discriptive names.

**Comment by dexX7** [comment]
> @zathras-crypto: just to be clear though, there are still quite a few bugs in the UI - but nothing that breaks things AFAIK

See https://github.com/zathras-crypto/omnicore/pull/70 for one that segfaults. ;)

But otherwise I agree, I didn't saw anything that was completely unexpected during the tests earlier (though I focused mostly on the cancel bug, and the general behavior after splitting tx 21).

File: src/qt/omnicore_init.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+{
+    // Don't skip, if disclaimer was requested explicitly
+    if (GetBoolArg("-disclaimer", false))
+        return true;
+
+    // Skip when starting minimized
+    if (GetBoolArg("-min", false))
+        return false;
+
+/* //TODO replace with 0.10 compatible mainnet check (see networkstyle)
```
Comment (dexX7) [inline src/qt/omnicore_init.cpp]: It's string based now:

```c++
    // Only mainnet users face the risk of monetary loss
    if (Params().NetworkIDString() != "main")
        return false;
```

File: src/mastercore_mdex.cpp [hunk 6]
```diff
@@ -398,8 +421,10 @@ int mastercore::MetaDEx_ADD(const std::string& sender_addr, uint32_t prop, int64
             return METADEX_ERROR -70;
         } else {
             // move tokens into reserve
-            assert(update_tally_map(sender_addr, prop, -new_mdex.getAmountRemaining(), BALANCE));
-            assert(update_tally_map(sender_addr, prop, new_mdex.getAmountRemaining(), METADEX_RESERVE));
+            //assert(update_tally_map(sender_addr, prop, -new_mdex.getAmountRemaining(), BALANCE));
```
Comment (dexX7) [inline src/mastercore_mdex.cpp]: Can you please re-enable those assertions?

File: src/qt/omnicore_init.cpp (hunk 5 above)
Comment (zathras-crypto) [inline src/qt/omnicore_init.cpp]: Thanks for the heads up - updated.

File: src/mastercore_mdex.cpp (hunk 6 above)
Comment (zathras-crypto) [inline src/mastercore_mdex.cpp]: Ahh yes sorry no longer needed to bypass them (I was bypassing them before your fixes went in to continue work on the UI).  Updated.

**Comment by dexX7** [comment]
I suggest to open a new PR for upcoming changes as soon as there are some. :)

**Pull request merged**